consensus/propagation/have_wants.go

```diff
@@ -85,118 +90,117 @@ func perPeerConcurrentRequestLimit() int64 {
 	return 3000
 }
 
-func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
-	var want *proptypes.WantParts
-	batchRequestCount := int64(0)
+func (blockProp *Reactor) requestFromPeer(ps *PeerState) {
 	for {
-		// if we're at a new height, we can drop the previous requests
-		if want != nil && !blockProp.relevant(want.Height, want.Round) {
-			want = nil
-			batchRequestCount = 0
-		}
+		availableReqs := ps.concurrentReqs.Load() + blockProp.ConcurrentRequestLimit()
 
-		// concurrent per-peer per-block request limit reached
-		if ps.requestCount.Load()+batchRequestCount >= perPeerConcurrentRequestLimit() {
-			err := blockProp.handleBatchLimit(ps, want)
-			if err != nil {
-				blockProp.Logger.Error("error handling batch limit", "err", err)
-			}
-			want = nil
-			batchRequestCount = 0
+		if availableReqs > 0 && len(ps.receivedHaves) > 0 {
+			ps.RequestsReady()
 		}
+
 		select {
 		case <-blockProp.ctx.Done():
 			return
-		case _, ok := <-ps.receivedPart:
-			if !ok {
-				return
-			}
-			ps.DecreaseRequestCount(1)
-		case req, ok := <-ps.requestChan:
+
+		case part, ok := <-ps.receivedParts:
 			if !ok {
 				return
 			}
-
-			partIndex := req.index
-			_, parts, fullReqs, has := blockProp.getAllState(req.height, req.round, false)
-			if !has {
-				blockProp.Logger.Error("couldn't find proposal when filtering requests", "height", req.height, "round", req.round)
-				continue
-			}
-			if parts.BitArray().GetIndex(int(partIndex)) {
+			if !blockProp.relevantHave(part.height, part.round) {
 				continue
 			}
+			ps.DecreaseConcurrentReqs(1)
 
-			reqLimit := 1
-			reqs := blockProp.countRequests(req.height, req.round, int(partIndex))
-			if len(reqs) >= reqLimit {
-				fullReqs.SetIndex(int(partIndex), true)
+		case <-ps.CanRequest():
+			canSend := blockProp.ConcurrentRequestLimit() - ps.concurrentReqs.Load()
+			if canSend <= 0 {
+				// should never be below zero
 				continue
 			}
 
-			// don't request the part from this peer if we've already requested it
-			// from them.
-			// this check will be helpful once we start requesting parts multiple times.
-			for _, p := range reqs {
-				// p == peer means we have already requested the part from this peer.
-				if p == ps.peer.ID() {
+			var (
+				wants    *proptypes.WantParts
+				parts    *proptypes.CombinedPartSet
+				fullReqs *bits.BitArray
+			)
+			for i := canSend; i > 0; {
```

Comment:
`i` doesn't seem to be decremented anywhere? if `canSend` is 1, and we have in the `receivedHaves` 10 requests, we will send them all of them even if the limit is 1

Reply:
whoops good find. removed i-- but forgot to add it later